File: ruby/lib/runapi/runway_aleph/resources/edit_video.rb
# frozen_string_literal: true

module RunApi
  module RunwayAleph
    module Resources
      # Runway Aleph video editing resource.
      # Transform an existing video using a text prompt and optional style reference image.
      class EditVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/runway_aleph/edit_video"
        RESPONSE_CLASS = Types::TaskCreateResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedEditVideoResponse
        MODEL = "runway-aleph"

        def initialize(http)
          @http = http
        end

        # Transform a video and wait until complete.
        #
        # @param params [Hash] edit parameters
        # @return [RunApi::RunwayAleph::Types::CompletedEditVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Start a video editing task.
        #
        # @param params [Hash] edit parameters
        # @return [RunApi::RunwayAleph::Types::TaskCreateResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["edit-video"], params.merge(model: MODEL))
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get video editing task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::RunwayAleph::Types::EditVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end

File: python/src/runapi/runway_aleph/resources/edit_video.py
"""Runway Aleph edit-video resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedEditVideoResponse,
    TaskCreateResponse,
)


class EditVideo(Resource):
    """Prompt-guided video editing with Runway Aleph."""

    ENDPOINT = "/api/v1/runway_aleph/edit_video"

    RESPONSE_CLASS = TaskCreateResponse
    COMPLETED_RESPONSE_CLASS = CompletedEditVideoResponse

    MODEL = "runway-aleph"

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a task and poll until it completes."""
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an edit-video task and return immediately with an ``id``."""
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["edit-video"], {**compacted, "model": self.MODEL})
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of an edit-video task."""
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
